src/library/arith_instance_manager.cpp

```diff
@@ -0,0 +1,489 @@
+/*
+Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: Daniel Selsam
+*/
+#include "util/sstream.h"
+#include "library/constants.h"
+#include "library/app_builder.h"
+#include "library/num.h"
+#include "library/util.h"
+#include "library/cache_helper.h"
+#include "library/arith_instance_manager.h"
+
+namespace lean {
+
+static arith_instance_info * g_nat_instance_info  = nullptr;
+static arith_instance_info * g_int_instance_info  = nullptr;
+static arith_instance_info * g_real_instance_info = nullptr;
+
+class arith_instance_info_cache {
+private:
+    environment m_env;
+    expr_struct_map<arith_instance_info> m_cache;
```

Comment:
Good point. I found the magic incantation to emplace the `arith_instance_info` on the map:
https://github.com/dselsam/lean/blob/smt2/src/library/arith_instance_manager.cpp#L484-L490